backend/czi_hosted/data_common/matrix_loader.py

```diff
@@ -133,7 +136,7 @@ def __init__(self, max_cached, timelimit_s=None):
         self.timelimit_s = timelimit_s
 
     @contextmanager
-    def data_adaptor(self, url_dataroot, location, app_config):
+    def data_adaptor(self, url_dataroot, location, app_config, dataset=None):
```

Comment:
All of the parameters combinations between app.py:get_data_adaptor() and this function are confusing, especially given the global (config) state used in some of the code branches.  The function signature seems to allow cases that aren't handled (eg, dataset==None and Portal API base URI is defined).

If I read the code right, you should never have a case where _both_ location and dataset are used and/or correct.  You might consider changing this function signature to more correctly match the function semantics.

My sense (with partial understanding) is that there are only three legal modes:
  1. single dataset, in which case the `location` is all you need
  2. multi-dataset, with Portal API used -- `dataset` is all you need
  3. multi-dataset, no Portal API -- `location` is all you need

Given this, the semantics are really a "redirect" based on location.  So you could change the function signature to something like the following and handle both:

    data_adaptor(self, url_dataroute, dataroot, dataset)

(and move the path_join from get_data_adaptor into this function)

Or, alternatively, create separate wrapper functions for both, and make the caller invoke the one with intended semantics?

Reply:
For case number 2 you still need the location to use as the key for the cache, otherwise the lookup will fail and we'll end up hitting the data portal endpoint for every request

Reply:
Point taken!   That still doesn't address the issue of how complex this particular function is and the number of independent routes through it (based on config changes).  It also spreads the logic of detering the S3 path across two separate modules (get_data_adaptor() uses `dataset` to make a physical path, and now data_adaptor() does the same thing with indirection via the poral).  Super likely this will be a source of future bugs IMHO.

Here is an alternative approach that (I think) solves both:
* do caching based on the `dataset` value
* add a new function which, given a `dataset`, returns two values:   the explorer path (ie, `join(dataroot, datapath)`, and the S3 dataset location.  Something like:   `getDatasetLocation(dataset:str) -> (explorer_path: str, S3_path: str)`
* All of the config-driven logic and communication with Portal are in that single function.  This function could even have its own cache to make things simpler (ie, don't rely on the adaptor cache)

This new function could be called by either get_data_adaptor or the matrix loader (assuming it has its own cache).  It makes the code in matrix_loader.py::data_adaptor()  more transparent, ie, you could define it as:

> def data_adaptor(self, base_url, explorer_dataset_url, S3_dataset_url)

And all of the config-related branching in data_adaptor() gets removed....